qutebrowser/misc/sql.py

```diff
@@ -168,75 +169,160 @@ def raise_sqlite_error(msg, error):
+#         # as those are idempotent, let's make sure we run them once again.
+#         Query("PRAGMA journal_mode=WAL").run()
+#         Query("PRAGMA synchronous=NORMAL").run()
+#
+#
+# def close():
+#     """Close the SQL connection."""
+#     QSqlDatabase.removeDatabase(QSqlDatabase.database().connectionName())
+#
+#
+# def version():
+#     """Return the sqlite version string."""
+#     try:
+#         if not QSqlDatabase.database().isOpen():
+#             init(':memory:')
+#             ver = Query("select sqlite_version()").run().value()
+#             close()
+#             return ver
+#         return Query("select sqlite_version()").run().value()
+#     except KnownError as e:
+#         return 'UNAVAILABLE ({})'.format(e)
+
+
+class Database:
+
+    """A wrapper over a QSqlDatabase connection."""
+
+    _USER_VERSION = UserVersion(0, 4)  # The current / newest user version
+
+    def __init__(self, path: str):
+        if QSqlDatabase.database(path).isValid():
+            raise BugError(f'A connection to the database at "{path}" already exists')
+
+        self._path = path
+        database = QSqlDatabase.addDatabase('QSQLITE', path)
+        if not database.isValid():
+            raise KnownError('Failed to add database. Are sqlite and Qt sqlite '
+                             'support installed?')
+        database.setDatabaseName(path)
+        if not database.open():
+            error = database.lastError()
+            msg = f"Failed to open sqlite database at {path}: {error.text()}"
+            raise_sqlite_error(msg, error)
+
+        version_int = self.query('pragma user_version').run().value()
+        self._user_version = UserVersion.from_int(version_int)
+
+        if self._user_version.major > self._USER_VERSION.major:
+            raise KnownError(
+                "Database is too new for this qutebrowser version (database version "
+                f"{self._user_version}, but {self._USER_VERSION.major}.x is supported)")
+
+        if self.user_version_changed():
+            log.sql.debug(f"Migrating from version {self._user_version} "
+                          f"to {self._USER_VERSION}")
+            # Note we're *not* updating _user_version here. We still want
+            # user_version_changed() to return True, as other modules (such as
+            # history.py) use it to create the initial table structure.
+            self.query(f'PRAGMA user_version = {_USER_VERSION.to_int()}').run()
+
+            # Enable write-ahead-logging and reduce disk write frequency
+            # see https://sqlite.org/pragma.html and issues #2930 and #3507
+            #
+            # We might already have done this (without a migration) in earlier versions,
+            # but as those are idempotent, let's make sure we run them once again.
+            self.query("PRAGMA journal_mode=WAL").run()
+            self.query("PRAGMA synchronous=NORMAL").run()
+
+    def qSqlDatabase(self):
```

Comment:
Had to make it public because `SqlQuery` needs to access it to start transactions.